Cabal/src/Distribution/Simple/Haddock.hs

```diff
@@ -800,6 +787,63 @@ getGhcLibDir verbosity lbi = do
     _ -> error "haddock only supports GHC and GHCJS"
   return $ mempty{argGhcLibDir = Flag l}
 
+-- | If Hi Haddock is supported, this function creates temporary directories
+-- and copies existing interface and object files produced by GHC into them,
+-- then passes them off to the given continuation.
+--
+-- If Hi Haddock is _not_ supported, we can't re-use GHC's compilation files.
+-- Instead, we use a clean temporary directory to the continuation,
+-- with no hope for recompilation avoidance.
+--
+-- See Note [Hi Haddock Recompilation Avoidance]
+reusingGHCCompilationArtifacts
+  :: Verbosity
+  -> TempFileOptions
+  -> Maybe (SymbolicPath CWD (Path.Dir Pkg))
+  -- ^ Working directory
+  -> LocalBuildInfo
+  -> BuildInfo
+  -> ComponentLocalBuildInfo
+  -> Version
+  -- ^ Haddock's version
+  -> ((SymbolicPath Pkg (Path.Dir Artifacts), SymbolicPath Pkg (Path.Dir Artifacts), SymbolicPath Pkg (Path.Dir Artifacts)) -> IO r)
+  -- ^ Continuation
+  -> IO r
+reusingGHCCompilationArtifacts verbosity tmpFileOpts mbWorkDir lbi bi clbi version act
+  | version >= mkVersion [2, 28, 0] = do
+      withTempDirectoryCwdEx verbosity tmpFileOpts mbWorkDir (distPrefLBI lbi) "haddock-objs" $ \tmpObjDir ->
+        withTempDirectoryCwdEx verbosity tmpFileOpts mbWorkDir (distPrefLBI lbi) "haddock-his" $ \tmpHiDir -> do
+          -- Re-use ghc's interface and obj files, but first copy them to
+          -- somewhere where it is safe if haddock overwrites them
+          let
+            vanillaOpts = componentGhcOptions normal lbi bi clbi (buildDir lbi)
+            i = interpretSymbolicPath mbWorkDir
+            copyDir ghcDir tmpDir = copyDirectoryRecursive verbosity (i $ fromFlag $ ghcDir vanillaOpts) (i tmpDir)
+          copyDir ghcOptObjDir tmpObjDir
+          copyDir ghcOptHiDir tmpHiDir
+          -- copyDir ghcOptStubDir tmpStubDir -- (see NB below)
+
+          act (tmpObjDir, tmpHiDir, fromFlag $ ghcOptHiDir vanillaOpts)
+
+  -- NB: Turns out, -stubdir is included in GHC's recompilation fingerprint.
+  -- This means that if we use a temporary directory for stubfiles
+  -- produced by GHC for the haddock invocation, haddock will trigger
+  -- full recompilation since the stubdir would be different.
+  --
+  -- So we don't use a temporary stubdir, despite the tmp o-dir and hi-dir:
+  --
+  -- We want to avoid at all costs haddock accidentally
+  -- overwriting o-files and hi-files (e.g. if a user specified
+  -- haddock-option triggers recompilation), and thus copy them to a
+  -- temporary directory to pass them on to haddock. However, stub files
+  -- are much less problematic since ABI-incompatibility isn't at play
+  -- here, that is, I don't think there exists a GHC flag that could
+  -- accidentally make a stub file incompatible with the one produced by
+  -- GHC from the same module.
```

Comment:
This is an interesting piece of detail. Shouldn't it live in the `Hi Haddock Recompilation Avoidance` note?

Reply:
I'll do that.